hub/core/lowlevel.py

```diff
@@ -0,0 +1,272 @@
+    return Pointer(c_array=(ctypes.c_byte * size)())
+
+
+def memcpy(dest: Pointer, src: Pointer, count=None) -> None:
+    if count is None:
+        count = src.size
+    ctypes.memmove(dest.address, src.address, count)
+
+
+def _write_pybytes(ptr: Pointer, byts: Union[bytes, memoryview]) -> Pointer:
+    memcpy(ptr, _ndarray_to_ptr(np.frombuffer(byts, dtype=np.byte)))
+    return ptr + len(byts)
+
+
+def _ndarray_to_ptr(arr: np.ndarray) -> Pointer:
+    return Pointer(arr.__array_interface__["data"][0], arr.itemsize * arr.size)
+
+
+def _pybytes_to_c_array(byts: bytes) -> Pointer:
+    return Pointer(
+        np.frombuffer(byts, dtype=np.byte).__array_interface__["data"][0], len(byts)
+    )
+
+
+def _infer_chunk_num_bytes(
+    version: str,
+    shape_info: np.ndarray,
+    byte_positions: np.ndarray,
+    data: Optional[Union[Sequence[bytes], Sequence[memoryview]]] = None,
+    len_data: Optional[int] = None,
+) -> int:
+    # NOTE: Assumption: version string contains ascii characters only (ord(c) < 128)
+    # NOTE: Assumption: len(version) < 256
+    assert len(version) < 256
+    assert max((map(ord, version))) < 128
+    # assert shape_info.ndim == 2
+    # assert byte_positions.ndim == 2
+    # version_slice_size = 1 + len(version)
+    # shape_info_slice_size = 4 + 4 + shape_info.nbytes
+    # byte_positions_slice_size = 4 + byte_positions.nbytes
+    # data_slice_size = sum(map(len, data))
+    if len_data is None:
+        len_data = sum(map(len, data))  # type: ignore
+    return len(version) + shape_info.nbytes + byte_positions.nbytes + len_data + 13
+
+
+def encode_chunk(
+    version: str,
+    shape_info: np.ndarray,
+    byte_positions: np.ndarray,
+    data: Union[Sequence[bytes], Sequence[memoryview]],
+    len_data: Optional[int] = None,
+) -> memoryview:
+
+    if len_data is None:
+        len_data = sum(map(len, data))
+
+    flatbuff = malloc(
+        _infer_chunk_num_bytes(version, shape_info, byte_positions, data, len_data)
+    )
+    ptr = flatbuff + 0
+
+    # write version
+    ptr[0] = len(version)
+    ptr += 1
+    for c in version:
+        ptr[0] = ord(c)
+        ptr += 1
+
+    # write shape info
+    ptr = _write_pybytes(ptr, np.int32(shape_info.shape[0]).tobytes())
+    ptr = _write_pybytes(ptr, np.int32(shape_info.shape[1]).tobytes())
+    memcpy(ptr, _ndarray_to_ptr(shape_info))
+    ptr += shape_info.nbytes
+
+    # write byte positions
+    ptr = _write_pybytes(ptr, np.int32(byte_positions.shape[0]).tobytes())
+    memcpy(ptr, _ndarray_to_ptr(byte_positions))
+    ptr += byte_positions.nbytes
+
+    # write actual data
+    for d in data:
+        if isinstance(d, Pointer):
+            d = d.memoryview
+        ptr = _write_pybytes(ptr, d)
+
+    return memoryview(flatbuff.bytes)
+
+
+def decode_chunk(
+    buff: Union[bytes, Pointer, memoryview]
+) -> Tuple[str, np.ndarray, np.ndarray, memoryview]:
+    if not isinstance(buff, Pointer):
+        buff = _pybytes_to_c_array(buff)
```

Comment:
also would be nice to have this in `hub.core.meta.encode.chunk`

(also docstring)